flang/lib/Optimizer/CodeGen/CodeGen.cpp

```diff
@@ -1468,22 +1497,27 @@ struct XArrayCoorOpConversion
     mlir::Value one = genConstantIndex(loc, idxTy, rewriter, 1);
     auto prevExt = one;
     mlir::Value off = genConstantIndex(loc, idxTy, rewriter, 0);
-    const bool isShifted = coor.shift().size() != 0;
-    const bool isSliced = coor.slice().size() != 0;
+    const bool isShifted = !coor.shift().empty();
+    const bool isSliced = !coor.slice().empty();
     for (unsigned i = 0; i < rank;
-         ++i, ++indexOps, ++shapeOps, ++shiftOps, ++sliceOps) {
+         ++i, ++indexOps, ++shapeOps, ++shiftOps, sliceOps += 3) {
       auto index = asType(loc, rewriter, idxTy, *indexOps);
       auto nextExt = asType(loc, rewriter, idxTy, *shapeOps);
       mlir::Value lb = one;
       if (isShifted)
         lb = asType(loc, rewriter, idxTy, *shiftOps);
       mlir::Value step = one;
-      if (isSliced)
-        step = asType(loc, rewriter, idxTy, *(sliceOps + 2));
+      auto normalSlice = isSliced;
+      if (isSliced) {
+        mlir::Value ub = *(sliceOps + 1);
+        normalSlice = !mlir::isa_and_nonnull<fir::UndefOp>(ub.getDefiningOp());
+        if (normalSlice)
+          step = asType(loc, rewriter, idxTy, *(sliceOps + 2));
+      }
```

Comment:
I did not see where the lower bound an "index" stride (not a normal stride) is applied. Is it taken into account somewhere as an offset ?